### PR TITLE
[tools/onert_train] Introduce onert_train tool

### PR DESCRIPTION
**Pull request by jyoungyun**
This commit introduces onert_train executable.
This tool has a train-related parameters such as Learning Rate, Optimizer, Loss function, and so on.

ONE-DCO-1.0-Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

Related issue: #10804

### CODE REVIEW

File: tests/tools/onert_train/src/onert_train.cc [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+#include <cassert>
+#include <chrono>
+#include <cstdlib>
+#include <iostream>
+#include <libgen.h>
+#include <stdexcept>
+#include <unordered_map>
+#include <vector>
+
+static const char *default_backend_cand = "training";
```
Comment (jyoungyun) [inline tests/tools/onert_train/src/onert_train.cc:36]: This part is not determined.

**Comment by glistening** [comment]
@jyoungyun Could you please use `[tool/onert_train]` or `[onert/tool] Introduce onert_train` or ...?

It is for specifying it is not about onert core, but about test or tool ...

File: tests/tools/onert_train/src/allocation.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace onert_train
+{
+class Allocation
+{
+public:
+  Allocation() : data_(nullptr) {}
+  ~Allocation() { free(data_); }
+  void *data() const { return data_; }
+  void *alloc(uint64_t sz) { return data_ = malloc(sz); }
```
Comment (ragmani) [inline tests/tools/onert_train/src/allocation.h:31]: It would be nice to add checking if `data_` has already been allocated. We have to ensure that there is no memory leak when this method is called more than once.
Or how about using `unique_ptr`?

**Review by ragmani — APPROVED** [review]
LGTM

**Comment by glistening** [comment]
@jyoungyun Please find https://github.com/Samsung/ONE/pull/10806#issuecomment-1562195476.

**Comment by jyoungyun** [comment]
I already changed the commit header but I didn't updated this issue title. So I fixed it.

**Comment by glistening** [comment]
@jyoungyun Could you leave what is the main modification or addition in this PR compared to `onert_run`? I cannot find arguments for `onert_train`. You may show how to use it in comment.

**Review by glistening — APPROVED** [review]
I guess it is introdution PR. The actual change will be done in following PRs.

**Comment by jyoungyun** [comment]
@glistening 

The arguments have been changed from `onert_run` code. And other codes are almost the same with the `onert_run`.

File: tests/tools/onert_train/src/args.cc [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    ("help,h", "Print available options")
+    ("version", "Print version and exit immediately")
+    ("nnpackage", po::value<std::string>()->notifier(process_nnpackage), "NN Package file(directory) name")
+    ("modelfile", po::value<std::string>()->notifier(process_modelfile), "NN Model filename")
+    ("path", po::value<std::string>()->notifier(process_path), "NN Package or NN Modelfile path")
+    ("epoch", po::value<int>()->default_value(5)->notifier([&](const auto &v) { _epoch = v; }), "Epoch number (default: 5)")
+    ("batch_size", po::value<int>()->default_value(32)->notifier([&](const auto &v) { _batch_size = v; }), "Batch size (default: 32)")
+    ("learning_rate", po::value<float>()->default_value(1.0e-4)->notifier([&](const auto &v) { _learning_rate = v; }), "Learning rate (default: 1.0e-4)")
+    ("loss", po::value<std::string>()->default_value("mse")->notifier([&] (const auto &v) { _loss_function = v; }), "Loss function name (default: mse)")
+    ("optimizer", po::value<std::string>()->default_value("sgd")->notifier([&] (const auto &v) { _optimizer = v; }), "Optimizer name (default: sgd)")
```
Comment (glistening) [inline tests/tools/onert_train/src/args.cc:191]: I found added args.
